core/kt-osrd-utils/src/main/kotlin/fr/sncf/osrd/utils/DistanceRangeMapImpl.kt

```diff
@@ -90,28 +99,54 @@ data class DistanceRangeMapImpl<T>(
             bounds.remove(0)
     }
 
+    /** Put a new bound and a matching value at the given offset,
+     * before the existing transition at that index if there is one */
+    private fun putTransitionBefore(offset: Distance, newValue: T?) {
+        var i = 0
+        while (i < bounds.size && bounds[i] < offset)
+            i++
+        bounds.insert(i, offset)
+        values.add(min(i, values.size), newValue)
+    }
+
+    /** Put a new bound and a matching value at the given offset,
+     * after the existing transition at that index if there is one */
+    private fun putTransitionAfter(offset: Distance, newValue: T?) {
+        var i = 0
+        while (i < bounds.size && bounds[i] <= offset)
+            i++
+        bounds.insert(i, offset)
+        values.add(min(i, values.size), newValue)
+    }
```

Comment:
I believe these functions can introduce duplicate bounds

Reply:
They do, but they're deleted afterwards. It's mostly the same behavior as before but cleared up.

It was surprisingly difficult to get it right.